internal/dag/gatewayapi_processor.go

```diff
@@ -155,13 +156,125 @@ func (p *GatewayAPIProcessor) Run(dag *DAG, source *KubernetesCache) {
 			}
 		}
 
+		// Validate that this route doesn't conflict with another.
+		matchingRoutes, rejectedRoutes := filterConflictingRoutes(matchingRoutes)
+
+		// Set status for all rejected routes.
+		for _, rejected := range rejectedRoutes {
+			//TODO: Make the 'message' descriptive to inform which resource is in conflict.
+			routeAccessor, commit := p.dag.StatusCache.ConditionsAccessor(k8s.NamespacedNameOf(rejected), rejected.Generation, status.ResourceHTTPRoute, rejected.Status.Gateways)
+			routeAccessor.AddCondition(gatewayapi_v1alpha1.ConditionRouteAdmitted, metav1.ConditionFalse, status.ReasonRouteConflict, "HTTPRoute rejected due to conflict with another.")
+			commit()
+		}
+
 		// Process all the routes that match this Gateway.
 		for _, matchingRoute := range matchingRoutes {
 			p.computeHTTPRoute(matchingRoute, listenerSecret)
 		}
 	}
 }
 
+type RouteMatch struct {
+	Object  *gatewayapi_v1alpha1.HTTPRoute
+	Path    *gatewayapi_v1alpha1.HTTPPathMatch
+	Headers *gatewayapi_v1alpha1.HTTPHeaderMatch
+}
+
+func filterConflictingRoutes(matchingRoutes []*gatewayapi_v1alpha1.HTTPRoute) (routesToProcess []*gatewayapi_v1alpha1.HTTPRoute, rejectedRoutes []*gatewayapi_v1alpha1.HTTPRoute) {
+
+	allMatches := make(map[string][]RouteMatch)
+	routesToProcess = copySlice(matchingRoutes)
+
+	// Gather up a list of matches by fqdn.
+	for _, route := range matchingRoutes {
+		for _, rule := range route.Spec.Rules {
+			for _, match := range rule.Matches {
+				for _, hostname := range route.Spec.Hostnames {
+
+					rm := RouteMatch{
+						Object:  route,
+						Path:    match.Path,
+						Headers: match.Headers,
+					}
+
+					// Check if this already exists in the set
+					for _, listMatch := range allMatches[string(hostname)] {
```

Comment:
does this end up being O(n^2)? we iterate every route -> hostname etc. and check if there is a conflict against any existing routes that have already been processed

i would argue for making route insertions on a vhost into the DAG smarter rather than trying to sort the valid routes before insertion since theyre already implemented as a map with the path, header, etc. serialized into the map key

see: https://github.com/projectcontour/contour/blob/73864b85309cd4cdb7c98abecc47996ecc0c8e58/internal/dag/dag.go#L485-L498

Reply:
Yup I agree with you but not sure the dag is the right place. The rules for which HTTPProxy wins is different than for which HTTPRoute wins, etc. 

I did it here to keep that generic layer (i.e. the dag) clean from any specific type.

Reply:
Let me look into your suggestion @sunjayBhatia  and see how it might work. 

@skriss do you have any thoughts as well?

Reply:
we could make it generic by adding a boolean field to the dag route that specifies if a route is "overridable" so there is no leaking types into the dag

ingress/httpproxy would set their routes as overridable but gateway routes would not be, so gateway routes could override the other resources on insertion, but gateway routes could not override each other

could also reverse the polarity of the field so the ingress/httpproxy code does not have to change, and only gateway api would set a "prevent override" to true